includes/AMP/Output_Buffer.php

```diff
@@ -0,0 +1,295 @@
+		if ( defined( '\AMP__VERSION' ) && version_compare( AMP__VERSION, WEBSTORIES_AMP_VERSION, '>=' ) ) {
+			return;
+		}
+
+		/*
+		 * Start output buffering at very low priority for sake of plugins and themes that use template_redirect
+		 * instead of template_include.
+		 */
+		$priority = defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : ~PHP_INT_MAX; // phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+		add_action( 'template_redirect', [ $this, 'start_output_buffering' ], $priority );
+
+		add_filter( 'web_stories_amp_sanitizers', [ $this, 'add_web_stories_amp_content_sanitizers' ] );
+	}
+
+	/**
+	 * Start output buffering.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::finish_output_buffering()
+	 *
+	 * @return void
+	 */
+	public function start_output_buffering() {
+		if ( ! is_singular( Story_Post_Type::POST_TYPE_SLUG ) ) {
+			return;
+		}
+
+		/*
+		 * Disable the New Relic Browser agent on AMP responses.
+		 * This prevents the New Relic from causing invalid AMP responses due the NREUM script it injects after the meta charset:
+		 * https://docs.newrelic.com/docs/browser/new-relic-browser/troubleshooting/google-amp-validator-fails-due-3rd-party-script
+		 * Sites with New Relic will need to specially configure New Relic for AMP:
+		 * https://docs.newrelic.com/docs/browser/new-relic-browser/installation/monitor-amp-pages-new-relic-browser
+		 */
+		if ( function_exists( 'newrelic_disable_autorum' ) ) {
+			newrelic_disable_autorum();
+		}
+
+		ob_start( [ $this, 'finish_output_buffering' ] );
+		$this->is_output_buffering = true;
+	}
+
+	/**
+	 * Determine whether output buffering has started.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::start_output_buffering()
+	 * @see Sanitization::finish_output_buffering()
+	 *
+	 * @return bool Whether output buffering has started.
+	 */
+	public function is_output_buffering(): bool {
+		return $this->is_output_buffering;
+	}
+
+	/**
+	 * Finish output buffering.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::start_output_buffering()
+	 *
+	 * @param string $response Buffered Response.
+	 * @return string Processed Response.
+	 */
+	public function finish_output_buffering( string $response ): string {
+		$this->is_output_buffering = false;
+
+		try {
+			$response = $this->prepare_response( $response );
+		} catch ( \Error $error ) { // Only PHP 7+.
+			$response = $this->render_error_page( $error );
+		} catch ( Exception $exception ) {
+			$response = $this->render_error_page( $exception );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Process response to ensure AMP validity.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param string $response HTML document response. By default it expects a complete document.
+	 * @return string AMP document response.
+	 */
+	public function prepare_response( string $response ): string {
+		// Enforce UTF-8 encoding as it is a requirement for AMP.
+		if ( ! headers_sent() ) {
+			header( 'Content-Type: text/html; charset=utf-8' );
+		}
```

Comment:
Can we ensure that all the headers that WordPress normally sends here are sent if we do this. [See](https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/wp-includes/class-wp.php#L391-L528). I want to make sure that cache control headers and any other custom headers, that developers may have added.

Reply:
I don't see how this would remove other headers sent by WordPress. Can you perhaps elaborate?

Reply:
Have you validated that by sending headers like this, that WordPress still sending, it's headers?

Reply:
Just tested and all the normal WordPress headers are there just fine